Vlan endpoint support in kata-runtime

kgunjikar opened the pull request:
Support for VlanEndpoint in kata runtime. 


```
-bash-4.2$ kubectl logs iperf-serverhigh1
------------------------------------------------------------
Server listening on TCP port 5001
TCP window size: 85.3 KByte (default)
------------------------------------------------------------
[  4] local 172.16.194.4 port 5001 connected with 172.16.194.5 port 38525
[  5] local 172.16.194.4 port 5001 connected with 172.16.194.5 port 38538
[  6] local 172.16.194.4 port 5001 connected with 172.16.194.5 port 38524
[  7] local 172.16.194.4 port 5001 connected with 172.16.194.5 port 38528
[  8] local 172.16.194.4 port 5001 connected with 172.16.194.5 port 38533
[  9] local 172.16.194.4 port 5001 connected with 172.16.194.5 port 38530
[ 10] local 172.16.194.4 port 5001 connected with 172.16.194.5 port 38532
[ 11] local 172.16.194.4 port 5001 connected with 172.16.194.5 port 38534
[ ID] Interval       Transfer     Bandwidth
[  4]  0.0-20.0 sec  2.38 GBytes  1.02 Gbits/sec
[  5]  0.0-20.0 sec  6.19 GBytes  2.66 Gbits/sec
[  7]  0.0-20.0 sec  2.66 GBytes  1.14 Gbits/sec
[  8]  0.0-20.0 sec  4.93 GBytes  2.12 Gbits/sec
[  9]  0.0-20.0 sec  1.46 GBytes   625 Mbits/sec
[ 10]  0.0-20.0 sec  1.30 GBytes   558 Mbits/sec
[ 11]  0.0-20.0 sec  1.13 GBytes   485 Mbits/sec
[  6]  0.0-20.0 sec  1.79 GBytes   769 Mbits/sec
[SUM]  0.0-20.0 sec  21.8 GBytes  9.37 Gbits/sec


-bash-4.2$ kubectl get pods
NAME                READY   STATUS    RESTARTS   AGE
iperf-clienthigh1   1/1     Running   0          22s
iperf-serverhigh1   1/1     Running   0          26s
-bash-4.2$ kubectl logs iperf-clienthigh1
WARNING: attempt to set TCP maximum segment size to 1454, but got 536
WARNING: attempt to set TCP maximum segment size to 1454, but got 536
WARNING: attempt to set TCP maximum segment size to 1454, but got 536
WARNING: attempt to set TCP maximum segment size to 1454, but got 536
WARNING: attempt to set TCP maximum segment size to 1454, but got 536
WARNING: attempt to set TCP maximum segment size to 1454, but got 536
WARNING: attempt to set TCP maximum segment size to 1454, but got 536
WARNING: attempt to set TCP maximum segment size to 1454, but got 536
------------------------------------------------------------
Client connecting to iperf-serverhigh1, TCP port 5001
TCP window size: 45.0 KByte (default)
------------------------------------------------------------
[  6] local 172.16.194.5 port 38525 connected with 172.16.194.4 port 5001
[  4] local 172.16.194.5 port 38524 connected with 172.16.194.4 port 5001
[  8] local 172.16.194.5 port 38530 connected with 172.16.194.4 port 5001
[  5] local 172.16.194.5 port 38533 connected with 172.16.194.4 port 5001
[ 10] local 172.16.194.5 port 38528 connected with 172.16.194.4 port 5001
[  9] local 172.16.194.5 port 38538 connected with 172.16.194.4 port 5001
[  7] local 172.16.194.5 port 38532 connected with 172.16.194.4 port 5001
[  3] local 172.16.194.5 port 38534 connected with 172.16.194.4 port 5001
[ ID] Interval       Transfer     Bandwidth
[  6]  0.0- 5.0 sec   488 MBytes   818 Mbits/sec
[  4]  0.0- 5.0 sec  1.33 GBytes  2.29 Gbits/sec
[  8]  0.0- 5.0 sec   543 MBytes   911 Mbits/sec
[ 10]  0.0- 5.0 sec   578 MBytes   970 Mbits/sec
[  9]  0.0- 5.0 sec  1.36 GBytes  2.34 Gbits/sec
[  3]  0.0- 5.0 sec   283 MBytes   474 Mbits/sec
[  5]  0.0- 5.0 sec   730 MBytes  1.22 Gbits/sec
[  7]  0.0- 5.0 sec   212 MBytes   355 Mbits/sec
[SUM]  0.0- 5.0 sec  5.46 GBytes  9.39 Gbits/sec
[  6]  5.0-10.0 sec   554 MBytes   930 Mbits/sec
[  8]  5.0-10.0 sec   336 MBytes   564 Mbits/sec
[  5]  5.0-10.0 sec  1.20 GBytes  2.06 Gbits/sec
[ 10]  5.0-10.0 sec   650 MBytes  1.09 Gbits/sec
[  9]  5.0-10.0 sec  1.61 GBytes  2.76 Gbits/sec
[  7]  5.0-10.0 sec   362 MBytes   607 Mbits/sec
[  3]  5.0-10.0 sec   334 MBytes   561 Mbits/sec


\-bash-4.2$ kubectl get pods
NAME                READY   STATUS    RESTARTS   AGE
iperf-clienthigh1   1/1     Running   0          2m26s
iperf-serverhigh1   1/1     Running   0          2m30s
-bash-4.2$ kubectl get pods -o wide
NAME                READY   STATUS    RESTARTS   AGE     IP             NODE       NOMINATED NODE   READINESS GATES
iperf-clienthigh1   1/1     Running   0          2m35s   172.16.194.5   ltsserv6   <none>           <none>
iperf-serverhigh1   1/1     Running   0          2m39s   172.16.194.4   ltsserv5   <none>           <none>
-bash-4.2$
```

Commit:
vlan endpoint support added

currently vlan endpoint support is not there. this pr is a step
towards vlan implementation to allow vlan based CNI plugins
with kata container runtime

Signed-off-by: Saurabh Deoras <saurabh@diamanti.com>

Please enter the commit message for your changes. Lines starting
with '#' will be ignored, and an empty message aborts the commit.

On branch 1.8.0-patched
Changes to be committed:
	modified:   virtcontainers/endpoint.go
	modified:   virtcontainers/network.go
	new file:   virtcontainers/vlan_endpoint.go

Commit:
Adding VlanEndPoint to the kata-runtime. There are few hacks wrt to UpdateRoutes but I'm able to bring up the container for now and pass traffic.

Events:
  Type    Reason     Age   From               Message
  ----    ------     ----  ----               -------
  Normal  Scheduled  54s   default-scheduler  Successfully assigned default/iperf-server-kata to ltsserv5
  Normal  Pulling    52s   kubelet, ltsserv5  Pulling image "diamanti/iperf:latest"
  Normal  Pulled     35s   kubelet, ltsserv5  Successfully pulled image "diamanti/iperf:latest"
  Normal  Created    34s   kubelet, ltsserv5  Created container iperf-server-kata
  Normal  Started    34s   kubelet, ltsserv5  Started container iperf-server-kata

Conflicts:
	virtcontainers/acrn_arch_base.go
	virtcontainers/persist.go
	virtcontainers/persist/api/network.go
	virtcontainers/qemu.go

sdeoras commented on virtcontainers/kata_agent.go:
```diff
@@ -532,25 +532,28 @@ func (k *kataAgent) updateInterfaces(interfaces []*vcTypes.Interface) error {
-		resultRoutes, ok := resultingRoutes.(*grpc.Routes)
-		if ok && resultRoutes != nil {
-			return k.convertToRoutes(resultRoutes.Routes), err
+	// XXX Remove me
```
just commenting to remind us in future that the commented code needs to be un-commented.

kgunjikar replied:
This will go away when we add metric to the route

sdeoras commented on virtcontainers/vlan_endpoint.go:
```diff
@@ -0,0 +1,117 @@
+
+// SetPciAddr sets the PCI address of the endpoint.
+func (endpoint *VlanEndpoint) SetPciAddr(pciAddr string) {
+	endpoint.PCIAddr = pciAddr
```
is there a pciAddr for vlan?

kgunjikar replied:
investigating, will remove if required

sdeoras reviewed (approved):
/lgtm
made some comments

Commit:
There is no support for metric in the Route construct, so for now I'm removing duplicate default route
so that the RouteAdd netlink call doesn't error out.

Metric support to be added .
